### DISCUSSION TITLE
List functions?

### DISCUSSION THREAD

**Discussion by bigkevmcd** [google/cel-go#1010] [Ideas]
I've implemented a couple of functions in projects that use CEL the easiest to link to is in Tekton Triggers:

https://github.com/tektoncd/triggers/blob/b133ef7b0d689c78a342879411a78c74b89de1f5/pkg/interceptors/cel/triggers.go#L202-L207

These are `.first()` and `.last()` which return the first and (unsurprisingly) last elements of a list.

This is useful when you want to split a string and take the first (or last) element in that list in a single expression, without having to resplit and calculate the length.

Would these be useful to add to the list extension?

I'm happy to do the work for this.

**Comment by TristonianJones** [comment]
Hi, I'd be totally happy to accept `first()` and `last()`. It's possible you'll want to use CEL's `OptionalTypes()` feature to indicate that the first and last might be absent in an empty list?

**Reply by TristonianJones** [reply]
Sorry for the delayed reply. I don't always see discussions right away.

**Reply by bigkevmcd** [reply]
I've got this working, but something that occurs to me, is the `cel.OptionalTypes()` might not be registered.

In this case...

```
lists_test.go:95: env.Check(![].first().hasValue()) failed: ERROR: <input>:1:21: undeclared reference to 'hasValue' (in container 'google.expr.proto2.test')
             | ![].first().hasValue()
             | ....................^
```

So, I added optional support for Optionals 😄
